src/collision/collide_and_slide.rs

```diff
@@ -0,0 +1,389 @@
+        origin: Vector,
+        mut velocity: Vector,
+        config: &CollideAndSlideConfig,
+        filter: &SpatialQueryFilter,
+        mut on_hit: impl FnMut(CollideAndSlideHitData) -> bool,
+    ) -> CollideAndSlideResult {
+        let mut position = origin;
+        let original_velocity = velocity;
+        let mut time_left = self.time.delta_secs();
+        let mut planes = config.planes.clone();
+
+        'outer: for _ in 0..config.collide_and_slide_iterations {
+            let sweep = time_left * velocity;
+            let Some((vel_dir, speed)) = Dir::new_and_length(sweep).ok() else {
+                // no more movement to go
+                break;
+            };
+            const MIN_SPEED: f32 = 0.0001;
+            if speed < MIN_SPEED {
+                break;
+            }
+
+            let depenetration_offset =
+                self.depenetrate(shape, shape_rotation, position, filter, config);
+            position += depenetration_offset;
+
+            let hit = self.query_pipeline.cast_shape(
+                shape,
+                position,
+                shape_rotation,
+                vel_dir,
+                &ShapeCastConfig::from_max_distance(speed),
+                filter,
+            );
+            let Some(hit) = hit else {
+                // moved the entire distance
+                position += sweep;
+                break;
+            };
+            let safe_distance = Self::pull_back(hit, vel_dir, config.skin_width);
+            if !on_hit(CollideAndSlideHitData {
+                hit,
+                position,
+                velocity,
+                safe_distance,
+            }) {
+                velocity = Vector::ZERO;
+                break 'outer;
+            }
+            if hit.distance == 0.0 {
+                // entity is completely trapped in another solid
+                velocity = Vector::ZERO;
+                break 'outer;
+            }
+            time_left -= time_left * (safe_distance / speed);
+
+            position += vel_dir * safe_distance;
+
+            // if this is the same plane we hit before, nudge velocity
+            // out along it, which fixes some epsilon issues with
+            // non-axial planes
+            let mut i = 0;
+            while i < planes.len() {
+                if hit.normal1.dot(planes[i].into()) > (1.0 - DOT_EPSILON) {
+                    velocity += hit.normal1 * config.duplicate_plane_nudge;
+                    break;
+                }
+                i += 1;
+            }
+            if i < planes.len() {
+                continue;
+            }
+            if planes.len() >= config.max_planes {
+                velocity = Vector::ZERO;
+                break 'outer;
+            }
+            planes.push(Dir::new_unchecked(hit.normal1));
+
+            // modify velocity so it parallels all of the clip planes
+
+            // find a plane that it enters
+            for i in 0..planes.len() {
```

Comment:
(comment applies to this entire `for in in 0..planes.len() {}` loop, but I couldn't convince github to let me select a span that long)

My (physicist's) mathematical intuition suggests this might be doable in a more robust way (hopefully one that doesn't need different implementation for 2d vs 3d) with a sprinkling of the Gram-Schmidt orthogonalisation procedure on the plane normals. It might need to be a slightly unusual version because the constraints are inequality conditions and not strict equality, but I'll see if I can figure out how to make it work if you're interested

Reply:
this one I will have to leave to you for sure hehe
But yes very open to alternatives!

Reply:
Sadly I haven't quite managed to firm up this idea into anything concrete enough to implement. I'll chime in again if I figure anything out I'm sure!